codegen/src/utils/fetch_metadata.rs

```diff
@@ -33,25 +34,62 @@ fn tokio_block_on<T, Fut: std::future::Future<Output = T>>(fut: Fut) -> T {
 
 /// Returns the metadata bytes from the provided URL.
 pub async fn fetch_metadata_bytes(url: &Uri) -> Result<Vec<u8>, FetchMetadataError> {
-    let hex = fetch_metadata_hex(url).await?;
-    let bytes = hex::decode(hex.trim_start_matches("0x"))?;
-    Ok(bytes)
-}
-
-/// Returns the raw, 0x prefixed metadata hex from the provided URL.
-pub async fn fetch_metadata_hex(url: &Uri) -> Result<String, FetchMetadataError> {
-    let hex_data = match url.scheme_str() {
+    let bytes = match url.scheme_str() {
         Some("http") | Some("https") => fetch_metadata_http(url).await,
         Some("ws") | Some("wss") => fetch_metadata_ws(url).await,
         invalid_scheme => {
             let scheme = invalid_scheme.unwrap_or("no scheme");
             Err(FetchMetadataError::InvalidScheme(scheme.to_owned()))
         }
     }?;
+
+    Ok(bytes)
+}
+
+/// Returns the raw, 0x prefixed metadata hex from the provided URL.
+pub async fn fetch_metadata_hex(url: &Uri) -> Result<String, FetchMetadataError> {
+    let bytes = fetch_metadata_bytes(url).await?;
+    let hex_data = format!("0x{}", hex::encode(bytes));
     Ok(hex_data)
 }
 
-async fn fetch_metadata_ws(url: &Uri) -> Result<String, FetchMetadataError> {
+/// Execute runtime API call and return the specified runtime metadata version.
+async fn fetch_latest_metadata(client: impl ClientT) -> Result<Vec<u8>, FetchMetadataError> {
+    const V15_METADATA_VERSION: u32 = u32::MAX;
+    let bytes = V15_METADATA_VERSION.encode();
```

Comment:
Given this this will be changing and won't necessarily be available on nodes or compatible with the `frame-metadata` that we'll try decoding against, I wonder whether we should only get the latest _stable_ metadata by default, and require users to call eg `--unstable-metadata` or something to ask for the unstable metadata?

I think I'd sleep better at night knowing that getting the unstable metadata is very much opt-in; it also means we won't be breaking the interface a bunch for "stable" users, while letting people experiment with runtime API call generation before it's stabilised.

(this also goes along with my desire to hide the unstable metadata we use internally I guess, though that is much more challenging to do in a nice way I feel!)

Can be done as a separate PR, but I think should be done :)